polygon/blocksync/block_downloader.go

```diff
@@ -0,0 +1,132 @@
+package blocksync
+
+import (
+	"context"
+	"errors"
+	"math/big"
+
+	"github.com/ledgerwatch/erigon/core/types"
+)
+
+func NewBlockDownloader(p2pLayer p2pLayer, dbLayer dbLayer, heimdallLayer heimdallLayer) *BlockDownloader {
+	return &BlockDownloader{
+		p2pLayer:      p2pLayer,
+		dbLayer:       dbLayer,
+		heimdallLayer: heimdallLayer,
+	}
+}
+
+type BlockDownloader struct {
+	p2pLayer      p2pLayer
+	dbLayer       dbLayer
+	heimdallLayer heimdallLayer
+}
+
+func (bd BlockDownloader) ForwardDownloadUsingCheckpoints(ctx context.Context, fromBlockNum uint64) error {
+	return bd.downloadUsingCheckpoints(ctx, fromBlockNum, false)
+}
+
+func (bd BlockDownloader) ForwardDownloadUsingMilestones(ctx context.Context, fromBlockNum uint64) error {
```

Comment:
I think that it should be a pointer receiver in all methods. I'll explain my reasoning in a meeting.

Reply:
addressed